includes/enrolment/class-sensei-course-enrolment-provider-state.php

```diff
@@ -0,0 +1,214 @@
+
+	/**
+	 * Log messages.
+	 *
+	 * @var array
+	 */
+	private $logs = [];
+
+	/**
+	 * Class constructor.
+	 *
+	 * @param Sensei_Course_Enrolment_Provider_State_Set $state_set        State set storing this provider state.
+	 * @param array                                      $provider_data    Basic storage for provider data.
+	 * @param array                                      $logs             Log messages.
+	 */
+	private function __construct( Sensei_Course_Enrolment_Provider_State_Set $state_set, $provider_data = [], $logs = [] ) {
+		$this->state_set     = $state_set;
+		$this->provider_data = $provider_data;
+		$this->logs          = $logs;
+	}
+
+	/**
+	 * Restore a course enrolment state record from data restored from a serialized JSON string.
+	 *
+	 * @param Sensei_Course_Enrolment_Provider_State_Set $state_set State set storing this provider state object.
+	 * @param array                                      $data      Serialized state of object.
+	 *
+	 * @return self|false
+	 */
+	public static function from_serialized_array( Sensei_Course_Enrolment_Provider_State_Set $state_set, $data ) {
+		if ( empty( $data ) ) {
+			return false;
+		}
+
+		$provider_data = isset( $data['d'] ) ? array_filter( array_map( [ __CLASS__, 'sanitize_data' ], $data['d'] ) ) : [];
+		$logs          = isset( $data['l'] ) ? array_filter( array_map( [ __CLASS__, 'sanitize_logs' ], $data['l'] ) ) : [];
+
+		return new self( $state_set, $provider_data, $logs );
+	}
+
+	/**
+	 * Create a fresh state storage record.
+	 *
+	 * @param Sensei_Course_Enrolment_Provider_State_Set $state_set State set storing this provider state object.
+	 *
+	 * @return self
+	 */
+	public static function create( Sensei_Course_Enrolment_Provider_State_Set $state_set ) {
+		return new self( $state_set );
+	}
+
+	/**
+	 * Sanitize a log entry.
+	 *
+	 * @param  array $log_entry Non-sanitized log entry.
+	 *
+	 * @return array
+	 */
+	private static function sanitize_logs( $log_entry ) {
+		if ( ! is_array( $log_entry ) ) {
+			return null;
+		}
+
+		if ( 2 !== count( $log_entry ) ) {
+			return null;
+		}
+
+		$log_entry[0] = intval( $log_entry[0] );
+		$log_entry[1] = sanitize_text_field( $log_entry[1] );
+
+		return $log_entry;
+	}
+
+	/**
+	 * Sanitize data value from provider storage.
+	 *
+	 * @param mixed $value Value to sanitize.
+	 *
+	 * @return mixed
+	 */
+	private static function sanitize_data( $value ) {
+		if ( is_array( $value ) ) {
+			return array_map( [ __CLASS__, 'sanitize_data' ], $value );
+		}
+
+		// These values are considered safe.
+		if ( is_int( $value ) || is_float( $value ) || is_bool( $value ) ) {
+			return $value;
+		}
+
+		// We don't allow object storage.
+		if ( is_object( $value ) ) {
+			return null;
+		}
+
+		return sanitize_text_field( $value );
+	}
+
+	/**
+	 * Return object that can be serialized by `json_encode()`.
+	 *
+	 * @return array
+	 */
+	public function jsonSerialize() {
+		return [
+			'd' => $this->provider_data,
```

Comment:
What's the reason for not using full words here?

Reply:
In its serialized form I thought it would be best to keep storage to a minimum. This object will be created for every user * every course. For most it will just be a simple flag for the manual provider, so I thought we might as well keep it smaller.